Update rust to 2024-08-24 nightly + remove `#[feature(asm_const)]`

hudson-ayers opened the pull request:
### Pull Request Overview

This pull request updates to the latest Rust nightly and removes the recently stabilized `#[feature(asm_const)`! It also reverts the litex riscv library to use the cleaner asm_const-based approach that we removed a while back when the bath to stabilization of asm_const was less clear.


### Testing Strategy

This pull request was tested by compiling/CI.


### TODO or Help Wanted

N/A

### Documentation Updated

- [x] no updates are required.

### Formatting

- [x] Ran `make prepush`.

lschuermann commented:
Oh awesome!

lschuermann commented:
This seems to elevate some of the warnings about `static mut` usages in our tests to errors. I already gave fixing these a shot as part of #4075. It is possible to avoid `static mut` entirely in the tests, however this is a non-trivial effort and requires refactoring of these tests (although it does not need `CoreLocal` of #3945).

Given that this PR does not block any other major feature, maybe it's a good idea to capitalize on its momentum and finally fix those tests for good.

hudson-ayers commented:
@lschuermann I fixed the linker script issue, and as part of investigating the test warnings I discovered that we are currently not failing the build on warnings in many cases -- we have 28 instances in the Makefile where we use `NOWARNINGS=true`, and that environment variable has no effect now that we moved away from the old make-based build system.

We still have a few tests that can (and do) use `RUSTFLAGS= -D warnings` because they do not include the `build.rs` script which forbids setting RUSTFLAGS directly, but large parts of Tock currently do not have forbid warnings on and several warnings have slipped into the build as a result.

I think that enabling warnings-as-errors again and fixing all those errors is probably a separate PR from this one since it is a bit tricky and probably somewhat large. I think it is fine if we want to wait to merge this PR until that PR happens though.

lschuermann commented:
Yes, I agree that this can go in first.

bradjc commented on boards/build.rs:
```diff
@@ -13,14 +13,21 @@
 use std::path::Path;
 
-const LINKER_SCRIPT: &str = "layout.ld";
+const LINKER_SCRIPT_OVERRIDE_ENV: &str = "LINKER_SCRIPT_OVERRIDE";
```
```suggestion
const LINKER_SCRIPT_OVERRIDE_ENV: &str = "TOCK_LINKER_SCRIPT";
```
?

bradjc commented on chips/stm32f4xx/src/clocks/clocks.rs:
```diff
@@ -621,6 +621,7 @@ impl<'a, ChipSpecs: ChipSpecsTrait> Stm32f4Clocks for Clocks<'a, ChipSpecs> {
 ///
 /// If there are any errors, open an issue ticket at <https://github.com/tock/tock>. Please provide the
 /// output of the test execution.
+#[allow(clippy::wildcard_imports)]
 pub mod tests {
     use super::*;
```
I don't think we should add this. I get that it's only for tests, but the same issues with `import *` are still there in that it is not clear what is needed for testing. This shouldn't have been merged in the first place.

bradjc commented on doc/OutOfTree.md:
```diff
@@ -119,7 +119,12 @@ to elements of Tock.
 Tock submodule's `build.rs` script to ensure it rebuilds when any linker scripts
-change.
+change. Alternatively if you have a linker script with a different name, you can
+set the `LINKER_SCRIPT_OVERRIDE` environment variable to pass the name of this
```
```suggestion
set the `TOCK_LINKER_SCRIPT` environment variable to pass the name of this
```

bradjc commented on rust-toolchain.toml:
```diff
@@ -3,6 +3,6 @@
 # Copyright Tock Contributors 2023.
 
 [toolchain]
-channel = "nightly-2024-07-08"
```
You need to run the update tock version script.

bradjc reviewed (changes requested):
Hooray, another nightly feature by the wayside!

I am concerned, however, with how unwieldy configuring Tock/Rust is. Determining what is actually building and how is now a combination of Cargo.toml, config.toml, Makefile, build.rs, and environment variables. I think part of the agreement with accepting hidden files was that those hidden files are _the defacto_ tool for configuring a cargo build. And yet we are looking to add a new prescribed way to configure how Tock is built.

I think we should not add the environment variable. Instead, boards should opt for a different build.rs file or we should use config.toml (somehow).

Also, I don't think we should add wildcard import exceptions.

Running the update rust script will update the version number in all locations it is stored.

hudson-ayers commented:
> I am concerned, however, with how unwieldy configuring Tock/Rust is. Determining what is actually building and how is now a combination of Cargo.toml, config.toml, Makefile, build.rs, and environment variables. I think part of the agreement with accepting hidden files was that those hidden files are _the defacto_ tool for configuring a cargo build. And yet we are looking to add a new prescribed way to configure how Tock is built.
> 
> 
> 
> I think we should not add the environment variable. Instead, boards should opt for a different build.rs file or we should use config.toml (somehow).

I will think on this a bit, I don't like duplicating the existing build.rs file or duplicating the whole board crate. Maybe there is something clever I can do here.


> Also, I don't think we should add wildcard import exceptions.

I think that these test modules are written following the standard rust conventions for tests, and just don't have `#[test]` annotations because they are intended as tests to run on actual hardware and that's not how our test infrastructure works today. I think https://github.com/rust-lang/rust-clippy/issues/5554 lays out the justification well for why clippy and the rust book encourage wildcard imports for this specific case only. 

I don't feel super strongly about this and it's easy to change back, but this allow was an intentional decision to follow rust conventions rather than a shortcut to get things working quickly. 

> Running the update rust script will update the version number in all locations it is stored.

Agh, I always forget this, will fix.

bradjc commented:
> I will think on this a bit, I don't like duplicating the existing build.rs file or duplicating the whole board crate. Maybe there is something clever I can do here.

I think we should create a build crate (like libtock-rs) for Tock, which would have the primary benefit of making it easier for out-of-tree boards to use the tock build.rs file. I [started](https://github.com/tock/tock/tree/buildrs-crate) on that, but it was more like an hour or two of work, not 15 minutes. It's possible that if we had that, this issue would be much simpler, but I'm not sure about that. If you think it would help, maybe we can get the build.rs crate done sooner rather than later.

> I think that these test modules are written following the standard rust conventions for tests, and just don't have `#[test]` annotations because they are intended as tests to run on actual hardware and that's not how our test infrastructure works today. I think [rust-lang/rust-clippy#5554](https://github.com/rust-lang/rust-clippy/issues/5554) lays out the justification well for why clippy and the rust book encourage wildcard imports for this specific case only.

Ok got it, I think you are right. Man, this is just the gift that keeps on giving. I think I end up leaving a comment on everything that stems from cargo's bad assumption because every single time it makes no sense in the context of Tock.

I wonder if there is a third party `#[test]`-like annotation for the cross-compiling use case we could use/advocate for.

ZhekaS commented:
Can this be broken down to separate PRs for removing the stabilized `#[feature(asm_const)]` and the rest? The former seem to be straightforward and will unblock building with stable rust (as long as the specific board does not use nightly cargo features)

bradjc commented:
We should be merging #4193 imminently which subsumes this PR.

ppannuto commented:
I just hit the merge button on #4193.
